### ISSUE TITLE
[RFC] Bring back simple sample-based playback

### ISSUE THREAD

**Issue by magcius** [WebAudio/web-audio-api#1322]
ScriptProcessorNode was deprecated without usable replacement. AudioWorkers and AudioWorklets are currently being implemented, I'm aware, but they are overcomplicated for the majority use case and still don't support runtime resampling, which is required for any complicated app that is streaming existing samples.

A lot of users of Web Audio are "shallow-graph" and are driven by an external audio loop which have constraints. I wrote about my frustrations with the Web Audio API [here](http://blog.mecheye.net/2017/09/i-dont-know-who-the-web-audio-api-is-designed-for/), and several people urged me to try to write a proposal on this issue tracker.

My proposal is to have a simple playback API, divorced from the graph-node-based-system. This is designed for people that want to have complete say and power over the exact samples being played, and is intended more as a spiritual successor to Mozilla's [Audio Data API](https://wiki.mozilla.org/Audio_Data_API). In my blog, I give a hypothetical example of such an API. An expanded version is here:

```js
const stream = new AudioStream(2, 32000);
const frequency = 440 / stream.sampleRate; // 440 Hz A note
stream.onfillsamples = function(samples) {
    const startTime = stream.startSamples; // The current "time", in samples.
    for (let i = 0; i < samples.length; i += 2) {
        const samp = startSamples + (i / 2); // Time, in samples.
        // samples is an Int16Array, interleaved PCM data.
        samples[i+0] = Math.sin(t * frequency) * 0x7FFF; // Channel 0 (L)
        samples[i+1] = Math.sin(t * frequency) * 0x7FFF; // Channel 1 (R)
    }
};
```

Details don't particularly matter to me, and I'm willing to take my time to write up a more formal specification for them, but I feel such an API would be a lot more valuable and tractable than ASBN, SPN, and the new AudioWorker / AudioWorklet, especially to consumers like emscripten.

Two important things:

1. There is *no* allocation mandated during the hot loop. I expect "samples" to be a slice of a ring buffer that the browser has internally.

2. The sample rate of the stream is provided by the user, similar to a buffer. We already allow resampling through ASBN, and most OS-level APIs allow resampling. That's a primary reason that emscripten uses ASBN instead of SPN.

If we are concerned about JS blocking the main thread, we can make this API accessible to WebWorker and ask users to use that if they are seeing dropping and hitching. If we are concerned about window.postMessage latency, we can ask users to communicate through a simple SharedArrayBuffer/futex scheme instead of slow message-passing.

Pinging @juj as he seems to be interested in such an API as well, for emscripten.

**Comment by juj** [comment]
For backstory, I have been working as the maintainer of Emscripten's audio backend engines (SDL, OpenAL) for the past 4-5 years, and during that time, have developed audio playback support to both. First we had Mozilla Audio Data API, then Web Audio AudioBufferSourceNodes, then briefly then Web Audio with ScriptProcessorNode, but shortly went back to Web Audio with ABSNs.

In native world, SDL API offers two ways to play back audio:
 - a pull-based streaming audio synthesis: SDL system periodically calls an user-provided audio callback in the context of a real-time audio pthread, which is expected to write out a fixed size buffer of samples to be played out from speakers shortly after.
 - a higher level full audio clip playback, with .wav, .ogg, .mp3 etc. support (depending on what system OS backend/compiled in codecs support)

In native world, OpenAL API likewise offers two ways to play back audio:
 - a push-based streaming audio synthesis: in C application main() code (browser main thread), one can keep pushing fixed or variable length blocks of raw audio PCM data, which the system queues back to back after each other to form a seamless audio stream. User can queue as much data up front in advance (unlike in pull-based method)
 - a higher level full audio clip playback, with raw PCM support (no compressed audio support)

All of these four use cases are critical, and we need to solve each of them:
 1. pull-based streaming synthesis
 2. push-based streaming synthesis
 3. raw PCM audio clip playback (8-bit/16-bit/24-bit, 8kHz-96kHz)
 4. compressed audio file playback
with various effects (looping with custom loop points, animating pitch, panning, possibility to hook in DSP effects), see e.g. https://github.com/juj/audio_test_suite

The case (1) is hard real-time, whereas (2) can be slightly less hard real-time (depending on application), and (3) and (4) don't have real-time thread timing requirements. One can emulate (1) with (2) and vice versa, but mixing push-pull boundaries is quite a troublesome thing to do unless real-time threading story of the platform is very good. (which it is not right now with wasm threads at least)

Mozilla Audio Data API was modelled after a push based synthesis approach (2), and it was a nice API layout-wise to work against. However because it tied in with DOM objects (`<audio>` element), it was not available in Workers, and on the main thread, JIT, GC, CC and tons of other browser scheduled effects completely prevented it from being suitable for anything remotely real time. Web Audio API's ScriptProcessorNode is somewhat identical to Mozilla Audio Data API, except that it was modelled in pull mode (1), and unfortunately had the exact same problems that Mozilla Audio Data API had.

I have no love lost for either API, they both attempted to solve a hard real time problem in a thread context where such real time guarantee is not available. For non-realtime uses of (2), Mozilla Audio Data API was fun to use, but it did not have a chance to work for real time use.

https://www.youtube.com/watch?v=boPEO2auJj4 is a good video about getting a grasp of "how realtime is realtime?", that is, how short buffer lengths people nowadays expect when they do realtime audio programming.

Discussing with the game development industry along the years, the trend is that the vast majority of audio using games care about (4) being solved efficiently, with the occassional effects (looping, animating pitch, panning, ...) tied in. The problems (3) and (4) are completely in the realm what Web Audio API could do, since these are not real-time problems, but it currently solves neither: support for (3) is missing due to issue #283, and support for (4) is missing due to issue #938. Only relatively few (but probably relatively more important as well!) professional applications, Unreal Engine 4 included (https://www.youtube.com/watch?v=ErejaBCicds) do the real-time critical bits in their own engine, and the rest would just be happy to play compressed audio files in a fire and forget manner, while setting the audio effects they want.

AudioWorklets+WebAssembly+SharedArrayBuffer intend to solve (1) proper. My understanding is that @padenot et al. have worked hard to design the API to be temporary garbage free, so that if used in a JS context that does not generate any garbage, there will be no GC or CC pauses. I am hopeful that it will be a good solution, but I am afraid about the JavaScript JIT, and how to write guaranteed JIT pause free code for (1) in an AudioWorklets context. Time will tell I think. Luke Wagner is heading a proposal to create "pure WebAssembly threads", which would be WebAssembly-executing threads that do not have an associated JS context in them at all. If nothing else, then eventually hooking audio APIs to such "pure wasm threads" should be able to finally solve the problem of providing a real-time audio thread context.

Currently Web Audio API looks like as if it had support for soft realtime (2). One can create several fixed length AudioBufferSourceNodes, and manually specify their playback times to lie back to back after each other, to produce a smooth continuous stream. This is what we currently do in Emscripten's SDL and OpenAL support for push and pull mode streaming synthesis. However, unfortunately this support falls short: Web Audio API does not guarantee anything about multiple buffers scheduled in such a manner, but even if a second audio buffer begins playback right after a first one ends, there is no guarantee that they will be seamlessly joined. A lot of work went into Firefox's Web Audio backend to ensure that the joins would be seamless in practice. If Web Audio API spec made this a behavior guaranteed by in the spec, then we could say that Web Audio API solves non-realtime portion of (2). I could not find an actual bug report about this, so marked this down as #1323. If that was mandated in the spec, then it would effectively give the same feature that Mozilla Audio Data API gave, except without #283.

I am waiting for AudioWorklets+WebAssembly+SharedArrayBuffer combo to land so that I can start prototyping. The intent is that we would then be able to implement (2) and (3) on top of multithreaded AudioWorklets. Then by compiling Ogg Vorbis and other codecs to WebAssembly, we'd also be able to offer support for (4) and bypass Web Audio API limitations on each of these. This will no doubt be quite a heavyweight solution, though perhaps it could be minified to a small "WebAudioPlayback API" library, which could offer (1)-(4) via an easy to include API. Ideally Web Audio API would be able to offer (1)-(4) directly in the core spec, so that all web developers would benefit without having to go to AudioWorklets+WebAssembly+SharedArrayBuffer, but perhaps a lot can be achieved here via a good npm library as well.

**Comment by hoch** [comment]
When I designed AudioWorklet, using it as a simple callback was clearly not the intention. So when I see all these wishes, it makes me a bit scared. AudioWorklet is not a magic trick or another escape hatch for WebAudio. I found two things in particular that might concern developers while I have been implementing it in Chrome:

1. WorkletGlobalScope is similar to WebWorker scope. Operations in the scope will create garbage no matter what. You can tweak the code to avoid GC as much as possible, but it will happen.

2. The thread for Worklet will have the "regular priority" because of the security restriction in the browser. This was not necessarily true when the rendering thread was hidden under the hood (using real-time priority), but to accommodate the idea of "exposing low-level audio access" the change was inevitable. It might perform badly on low end machines according to our recent investigation. (due to priority inversion + deschduled task)

It's going to be exciting to see AudioWorklet, WASM and SAB in play, but please keep in mind that was not the primary use case we envisioned. From my perspective, AudioWorklet is bringing the concept of "extensible web" to Web Audio API.

**Comment by magcius** [comment]
We're aware that the web has difficult scheduling constraints and that Web Workers and Worklets aren't a magic bullet that will solve all of our problems. That said, it seems to be the most anticipated building block from a lot of people's perspective, since it allows us to do sample-level synthesis and playback without being blocked by the UI / render thread.

I'm curious why you think GC is "guaranteed" to happen in a Worklet's scope. From what I can tell, a WorkletGlobalScope is supposed to be stateless, and be thread-independent. The only place where garbage can pile up is on the class itself (according to [0]), but as long as you don't make allocations in the hot path, aren't you going to be garbage-free?

[0] https://www.w3.org/TR/worklets-1/#code-idempotency

**Comment by hoch** [comment]
If you only think about AW/WASM/SAB use case, that might work okay. Because you create 1 AWN/AWP pair and 1 SAB and transfer the buffer to some other worker thread running WASM code. In this specific case, you only creates 2 objects in Worklet scope and you'll be using it until the tab is closed.

However, I do also care about other "regular" use cases: AWN and AWP are supposed to be used like any other AudioNode. You should be able to create multiple instances at any time - and the created AWP instances means garbages in the Worklet scope/thread. Perhaps it is worthy of non-normative guidance (or an explainer) in the spec.

**Comment by magcius** [comment]
Assuming I care about one custom AudioWorklet connected directly to graph output, but I'm not using WASM/SAB, simply JS in the Worklet thread, then I should be safe assuming I set up all the objects I need at construct time, correct?

I suppose there could be issues if I try to call methods / set properties on the Worklet, since I think the resulting WorkletScope will own the message contents, but I expect that to be relatively infrequent in practice.

**Comment by hoch** [comment]
`postMessage()` and `onmessage` on AWN/AWP do cross-thread data transaction via structured-serialization. (i.e. copying). Thus it creates garbage.

The frequency of transaction completely depends on the application. Many WebAudio use cases update the AudioParam setter inside of animation loop. That's every ~16ms.

**Comment by magcius** [comment]
It would be interesting to see a faster, garbage-less parameter update system -- as I mention in the original post, one could be built using SAB, where we treat it as a parameter buffer: main thread writes, AudioWorklet reads. It's a bit unconventional, but it does limit the garbage to be a fixed-size buffer per AudioWorklet node.

If we think this is edge case behavior and we expect most clients to update fairly infrequently, then I don't mind if other clients need special code for the high-performance case.

**Comment by juj** [comment]
> When I designed AudioWorklet, using it as a simple callback was clearly not the intention. So when I see all these wishes, it makes me a bit scared.

> It's going to be exciting to see AudioWorklet, WASM and SAB in play, but please keep in mind that was not the primary use case we envisioned.

Hmm, reading this, I notice that my exposure to AudioWorklets comes from these two pages alone: [AudioWorklet IDL Proposal](https://github.com/WebAudio/web-audio-api/wiki/AudioWorklet-IDL-Proposal) and [AudioWorklet Examples](https://github.com/WebAudio/web-audio-api/wiki/AudioWorklet-Examples) and I realize that is not much. To clarify the scope, what is the "poster problem" that AudioWorklets are intended to solve? Is there something with WASM and SAB specifically that causes this worry you mention?

> Two important things:
> There is no allocation mandated during the hot loop.

> It would be interesting to see a faster, garbage-less parameter update system

I think this conversation might be going down a fallacious trail of thought: "since GC pauses can cause stalls, if I can ensure there are no GC pauses, then there must be no stalls either". Opened issue #1327 to further discuss this aspect.

**Comment by hoch** [comment]
> AudioWorklet IDL Proposal and AudioWorklet Examples and I realize that is not much.

Because those were early discussions on the facade of API mostly. Perhaps reading this section might help: https://webaudio.github.io/web-audio-api/#AudioWorklet

> To clarify the scope, what is the "poster problem" that AudioWorklets are intended to solve?

I have been very clear about this: to provide the lowest primitive of audio processing within WebAudio ecosystem and to improve the extensibility of Web Audio API. The biggest motivation by far was the flawed design of ScriptProcessorNode.

> Is there something with WASM and SAB specifically that causes this worry you mention?

There is no "wrong way" to use the feature in the API. However, considering the idea of WASM+SAB came much later than the design of AudioWorklet, I don't think we thoroughly explored the idea. I was actually against the idea of supporting postMessage/Transferrable because of its complexity.

> I think this conversation might be going down a fallacious trail of thought

I don't think it is necessarily a "fallacious trail of thought". Avoiding GC would be one of the best practices in using AudioWorklet. Load balancing the JS code to avoid underflow (or missed deadline) is also critical and that is developer's responsibility. UA's job is to make sure it does not generate unnecessary garbages in the process, but it cannot guarantee anything when the user-supplied code passes the render quantum deadline.

**Comment by joeberkovitz** [comment]
WG discussed today, and we think this should be deferred to post V1 because:

- SAB, wherever available, will be supported in AudioWorklets out of the box, and eliminates any messaging-related garbage generation.

- Any alternative low-level approach is going to take a fair bit of effort to agree on, and determining the necessity of it is will depend on what emerges following the launch of AudioWorklets.

**Comment by joeberkovitz** [comment]
Just noting that we're talking directly with the WebAssembly folks to get a really tight integration with no GC. This will not be an instant fix but will require some thought and perhaps the WASM/WebAudio connection will look different from JS/WebAudio.

**Comment by joeberkovitz** [comment]
Also note #1443

**Comment by joeberkovitz** [comment]
Here's another way of looking at the "push" issue (for those who don't like or don't want the complexity of a WASM+SAB+AudioWorklet approach):

What about a shrink-wrapped FIFO-type source node -- call it, say, `StreamingBufferSourceNode` -- that enqueues AudioBuffers, and delivers them to its output seamlessly spliced together? You can push as much audio into it as you want, and it will eventually spit it out into the graph. One could also ask it how many sample frames are queued up inside it to decide if it needs refilling.

Could also dispatch events on hitting a low-water or high-water mark... obviously variations are possible.

Thoughts?

**Comment by magcius** [comment]
Something that would be convenient is a safe way to reuse AudioBuffers, or to create AudioBuffers out of slices of larger buffers without copying. Currently, creating AudioBuffers also means creating the storage inside them as well, which creates GC pressure, especially for short buffers.

**Comment by joeberkovitz** [comment]
Thanks, but we're specifically asking here for feedback on the idea of a node that cleanly supports push-style audio generation.

Reuse of AudioBuffers is a separate issue, which you can file -- but take note that AudioBuffers work the way they do because of the need to avoid data races in the Web platform. We've had to avoid many things that seem convenient at first blush, but have the potential to produce very unpredictable behavior.

**Comment by magcius** [comment]
Unfortunately, I cannot comment about that proposal in a vacuum. I have concerns about AudioBuffer GC waste, assuming your proposal means that I would have to create a new AudioBuffer at each push.

**Comment by joeberkovitz** [comment]
OK. Conservatively, for the purposes of discussion, assume that you have to create new AudioBuffers. Does a node that enqueues buffers seamlessly for playback (and offers built in resampling) improve on using a sequence of serially scheduled AudioBufferSourceNodes?

This is not an attempt to set aside the AudioBuffer question, just trying to clearly separate the issues. We can handle tracking the two issues' relationship.

**Comment by magcius** [comment]
For my use case, no. ABSN already has built-in resampling and a guarantee that playback is seamless.

**Comment by padenot** [comment]
This now possible and stable in the spec:

- Have a `WebWorker` that generates the audio (or the main thread, but well...)
- Have an `AudioWorklet`
- Have a `SharedArrayBuffer` shared between the two
- Use this `SharedArrayBuffer` as an spsc ring buffer: the `WebWorker` produces the audio, puts it in the queue. Whenever the `AudioWorklet` needs to process some audio, pull some data from the `spsc` queue and copy it to its output.

There is no garbage generated, and the setup is minimal.

Thoughts?

**Comment by hoch** [comment]
With primitives laid out by @padenot above, I created a proof-of-concept example based on them:
https://developers.google.com/web/updates/2018/06/audio-worklet-design-pattern#webaudio_powerhouse_audio_worklet_and_sharedarraybuffer

It seems complicated, but a lot of parts can be abstracted as a form of library. 
We're closing the issue, but feel free to re-open it if necessary.
